Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm

```diff
@@ -0,0 +1,188 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "config.h"
+#import "WebExtensionContext.h"
+
+#import "APIContentWorld.h"
+#import "CocoaHelpers.h"
+#import "WKContentWorld.h"
+#import "WKFrameInfoPrivate.h"
+#import "WKWebViewPrivate.h"
+#import "WebExtensionContextProxy.h"
+#import "WebExtensionContextProxyMessages.h"
+#import "WebExtensionScriptInjectionParameters.h"
+#import "WebExtensionScriptInjectionResultParameters.h"
+#import "WebExtensionDynamicScripts.h"
+#import "WebExtensionTab.h"
+#import "WebExtensionTabIdentifier.h"
+#import "WebExtensionUtilities.h"
+#import "WebPageProxy.h"
+#import "WebProcessProxy.h"
+#import "WebUserContentControllerProxy.h"
+#import "_WKFrameHandle.h"
+#import "_WKFrameTreeNode.h"
+#import <WebCore/UserStyleSheet.h>
+#import <wtf/BlockPtr.h>
+#import <wtf/cocoa/VectorCocoa.h>
+
+namespace WebKit {
+using namespace DynamicScripts;
+
+using SourcePair = DynamicScripts::SourcePair;
+using ScriptVector = DynamicScripts::StyleSheetPairs;
+
+void WebExtensionContext::scriptingExecuteScript(const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void(std::optional<Vector<WebExtensionScriptInjectionResultParameters>>)>&& callback)
+{
+    WKWebView *webView = m_tabMap.get(parameters.identifier.value())->mainWebView();
+    if (!hasPermission(webView.URL)) {
+        callback(std::nullopt);
+        return;
+    }
+
+    [webView _frames:makeBlockPtr([this, protectedThis = Ref { *this }, webView = RetainPtr { webView }, parameters = WTFMove(parameters), callback = WTFMove(callback)](_WKFrameTreeNode *mainFrame) mutable {
+        API::ContentWorld& executionWorld = *m_contentScriptWorld;
+        if (parameters.world == WebExtensionContentWorldType::Main)
+            executionWorld = API::ContentWorld::pageContentWorld();
+
+        Vector<RetainPtr<_WKFrameTreeNode>> frames;
+        if (parameters.allFrames || !parameters.frameIDs)
```

Comment:
You might just be able to drop `allFrames` from parameters and use `!parameters.frameIDs` to mean all frames, that avoids the issues where they could be out off sync with each other.